deploy/tests/test_generate_production_ini.py

```diff
@@ -81,39 +110,288 @@ def test_build_ini_file_from_template():
             else:
                 assert mode == 'w'
                 assert filename == some_ini_file_name
                 return MockFileStream(filename, mode)
 
-        with mock.patch("io.open", side_effect=mocked_open):
-            build_ini_file_from_template(some_template_file_name, some_ini_file_name)
+        with mock.patch("subprocess.check_output") as mock_check_output:
+            mock_check_output.side_effect = make_mocked_check_output_for_get_version()
+            with mock.patch("os.path.exists") as mock_exists:
+                def mocked_exists(filename):
+                    return filename in [EB_MANIFEST_FILENAME, some_template_file_name]
+                mock_exists.side_effect = mocked_exists
+                with mock.patch("io.open", side_effect=mocked_open):
+                    build_ini_file_from_template(some_template_file_name, some_ini_file_name)
+
+        assert MockFileStream.FILE_SYSTEM[some_ini_file_name] == [
+            '[Foo]',
+            'DATABASE = "snow_white"',
+            'SOME_URL = "http://user@unittest:6543/"',
+            'OOPS = "$NOT_AN_ENV_VAR"',
+            'HMMM = "${NOT_AN_ENV_VAR_EITHER}"',
+            'SHHH = "my-secret"',
+            'VERSION = "%s"' % MOCKED_BUNDLE_VERSION,
+            'PROJECT_VERSION = "%s"' % MOCKED_PROJECT_VERSION,
+        ]
+
+        MockFileStream.reset()
 
-        assert lines == [
+        with mock.patch("subprocess.check_output") as mock_check_output:
+            mock_check_output.side_effect = make_mocked_check_output_for_get_version()
+            with mock.patch("os.path.exists") as mock_exists:
+                def mocked_exists(filename):
+                    # Important to this test: This will return False for EB_MANIFEST_FILENAME,
+                    # causing the strategy of using the version there to fall through,
+                    # so we expect to try using the git version instead.
+                    return filename in [some_template_file_name]
+                mock_exists.side_effect = mocked_exists
+                with mock.patch("io.open", side_effect=mocked_open):
+                    build_ini_file_from_template(some_template_file_name, some_ini_file_name)
+
+        assert MockFileStream.FILE_SYSTEM[some_ini_file_name] == [
             '[Foo]',
             'DATABASE = "snow_white"',
             'SOME_URL = "http://user@unittest:6543/"',
             'OOPS = "$NOT_AN_ENV_VAR"',
             'HMMM = "${NOT_AN_ENV_VAR_EITHER}"',
             'SHHH = "my-secret"',
+            'VERSION = "%s"' % MOCKED_LOCAL_GIT_VERSION,
+            'PROJECT_VERSION = "%s"' % MOCKED_PROJECT_VERSION,
         ]
+
+        MockFileStream.reset()
+
+        with mock.patch("subprocess.check_output") as mock_check_output:
+            mock_check_output.side_effect = make_mocked_check_output_for_get_version(simulate_git_command=False)
+            with mock.patch("os.path.exists") as mock_exists:
+                def mocked_exists(filename):
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [function already defined line 161](https://app.codacy.com/manual/4dn/fourfront/pullRequest?prid=5434981)

Reply:
These are internal functions within tests and this is neither an actual nor a style error.
We really need to fix this diagnostic to ignore internal definitions, especially in test files.